Additional measurement support for Clifford device

obliviateandsurrender opened the pull request:
**Context:** Adds missing measurement supports for the Clifford device

**Description of the Change:** This PR adds support for the missing measurement functionalities for the Clifford device. This includes the following:

1. Analytical and statistical `expectation value` and `variance` measurements with all PennyLane observables, including the arithmetic operators. 
2. Support for all quantum information-related observables, including `purity` and `entropy`-based quantities.
3. Support for all sampling-based quantities such as `counts`, `samples`, etc. 
4. Support for the `classical shadow` based measurement routines.
5. Support for providing observables with different measurement operations. 

**Benefits:** Clifford device would support all PL-provided measurements. 

**Possible Drawbacks:** Some measurement routines, such as classical shadow-based methods might not be optimal. 

**Related GitHub Issues:** N/A

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -466,50 +478,256 @@ def simulate(
+            return coeffs, paulis
+
+        # Case for higher arithmetic depth for prod-type observables
+        if meas_obs.arithmetic_depth > 1 and isinstance(meas_obs, (Prod, SProd)):
+            check_op_math = qml.operation.active_new_opmath()
+            qml.operation.enable_new_opmath()
+            meas_obs_simp = meas_obs.simplify()
+            if not check_op_math:
+                qml.operation.disable_new_opmath()
```
I _think_ there should be a more elegant way to do this. I don't know what it is off the top of my head, but I'd like to confirm with an expert. I'm also curious if my recent bugfixes circumvent the need for this. Can we try removing it and re-running tests? If it's still a hassle, then nbd

obliviateandsurrender replied:
There might be a more elegant way, but this was the most straightforward one that came to me. This specific `if` branch deals with the case where I have an object built using nested `Prod` and `SProd`.

obliviateandsurrender replied:
I was able to reduce the logic for the old operators, but I would be keeping the logic for new ones. :)

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -466,50 +478,256 @@ def simulate(
+            coeffs.append(1.0)
+            paulis.append(
+                (
+                    "".join([_GATE_OPERATIONS[op.name] for op in meas_obs.terms()[1][0]]),
```
not sure what exactly is going on here, but I believe `Prod.terms()` is trivial - I think this is the same as `meas_obs[0]`? same comment below

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -466,50 +478,256 @@ def simulate(
+            paulis.append(("".join([_GATE_OPERATIONS[op.name] for op in gate_ops]), meas_obs.wires))
+            return coeffs, paulis
+
+        # Add support for more case when the time is right
```
please make sure to add a story to track any future work that is needed/nice to have:)

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -466,50 +478,256 @@ def simulate(
-                if isinstance(meas, DensityMatrixMP):  # do first because it is a child of StateMP
-                    res = self._measure_density_matrix(tableau_simulator)
+            # Computing samples via stim's compiled sampler
+            if isinstance(meas, SampleMP):
+                results.append(
+                    self._measure_observable_sample(meas.obs, stim_circuit, num_shots, sample_seed)[
```
do we only support sampling observables, eg. if a user does `qml.sample(qml.PauliZ(0))` as opposed to `qml.sample(wires=[0])`?

obliviateandsurrender replied:
Ahh, added it. :)

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -534,46 +752,233 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+
+    def _measure_variance(self, tableau_simulator, meas_op, stim):
+        """Measure the variance with respect to the state of simulator device."""
+        # TODO: find a better pythonic way to deal with this enable-disable
```
😄 if it's desirable we could consider adding a contextmanager type of thing

obliviateandsurrender replied:
I added it. Works like a beauty. I can make it private if you don't want it to be public.

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -534,46 +752,233 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+
+        if self._prob_states is None and self._tableau:
+            raise ValueError(
+                "In order to maintain computational efficiency,\
+                with ``tableau=True``, the clifford device supports returning\
+                probability only for selected target computational basis states.\
+                Please use `probability_target()` method to set them."
```
```suggestion
                "In order to maintain computational efficiency, \
                with ``tableau=True``, the clifford device supports returning \
                probability only for selected target computational basis states. \
                Please use `probability_target()` method to set them."
```

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -534,46 +752,233 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+                Please use `probability_target()` method to set them."
+            )
+
+        # we might be able to skip the inverse done below
+        # as the distribution for the inverted tableau should remain same
```
can we?

obliviateandsurrender replied:
Yes, I will update after testing it a bit more.

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -534,46 +752,233 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+        # as the distribution for the inverted tableau should remain same
+        diagonalizing_cit = tableau_simulator.current_inverse_tableau().inverse().to_circuit()
+        diagonalizing_ops = [] if not meas_op.obs else meas_op.obs.diagonalizing_gates()
+        for ops in diagonalizing_ops:
+            # Check if it is Clifford
+            if ops.name not in _GATE_OPERATIONS:  # pragma: no cover
+                raise ValueError(
+                    f"Currently, we only support observables whose diagonalizing gates are Clifford, got {ops}"
+                )
+            # Add to the circuit to rotate the basis
+            stim_op = self.pl_to_stim(ops)
```
nitpick (unless you're using it elsewhere?)
```suggestion
        for op in diagonalizing_ops:
            # Check if it is Clifford
            if op.name not in _GATE_OPERATIONS:  # pragma: no cover
                raise ValueError(
                    f"Currently, we only support observables whose diagonalizing gates are Clifford, got {op}"
                )
            # Add to the circuit to rotate the basis
            stim_op = self.pl_to_stim(op)
```

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -534,46 +752,233 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+            diagonalizing_sim.do_circuit(diagonalizing_cit)
+            for idx, wire in enumerate(meas_wires):
+                outcome = tgt_states[ind][idx]
+                expectaction = diagonalizing_sim.peek_z(wire)
+                if not expectaction:
+                    prob_res[ind] /= 2.0
+                elif expectaction != (-1 if outcome else 1):
```
```suggestion
                expectation = diagonalizing_sim.peek_z(wire)
                if not expectation:
                    prob_res[ind] /= 2.0
                elif expectation != (-1 if outcome else 1):
```

timmysilv reviewed (commented):
looks good so far! will review tests tomorrow

timmysilv reviewed (commented):
tests look good!

trbromley commented on pennylane/operation.py:
```diff
@@ -3017,6 +3018,22 @@ def convert_to_opmath(op):
     return op
 
 
+@contextmanager
+def use_new_opmath():
+    r"""Allows to use the new arithmetic operator dunders within a
+    temporary context using the `with` statement."""
+
+    check_op_math = active_new_opmath()
+    try:
+        enable_new_opmath()
+        yield
+    except Exception as e:
+        raise e
+    finally:
+        if not check_op_math:
+            disable_new_opmath()
+
```
Something we can align with @Qottmann and @mudit2812 on. We do plan for op arithmetic to be enabled by default in 0.36, so this context manager could still serve a purpose in 0.35.

Qottmann replied:
Is there an advantage to use this over just
```python
active_new_opmath()
# code
disable_new_optmath()
```
or just extra style points 😎 ?

obliviateandsurrender replied:
If the user had enabled new operator arithmetic for their environment, then simply calling `disable_new_opmath()` might cause a side effect and turn it off for them. One way to deal with this would be to check if it was initially active before calling the disabling method.

``` py
active_flag = active_new_opmath()
enable_new_opmath()
# code
if not active_flag:
    disable_new_opmath()
```

A cleaner (and pythonic) way of doing this would be to use context manager so that the toggling of operator arithmetic happens within that context block (defined via `with`) and prevents any side effect beyond it -

``` py
with use_new_opmath():
    #code
```

trbromley commented on tests/devices/test_default_clifford.py:
```diff
@@ -106,26 +106,219 @@ def circuit_empty():
     assert np.allclose(qnode_clfrd(), qnode_qubit())
 
 
+@pytest.mark.parametrize("shots", [None, int(1e6)])
+@pytest.mark.parametrize(
+    "ops",
+    [
+        qml.PauliX(0),
+        qml.PauliX(0) @ qml.PauliY(1),
+        qml.Hamiltonian([3.0, 2.0], [qml.PauliX(0), qml.PauliY(1)]),
+        qml.Hamiltonian([0.42], [qml.PauliZ(0) @ qml.PauliX(1)]),
+        qml.sum(qml.PauliX(0), qml.PauliY(1)),
+        qml.prod(qml.PauliX(0), qml.PauliY(1)),
+        qml.s_prod(3.0, qml.PauliX(0)),
+        qml.sum(qml.PauliX(0), qml.s_prod(2.0, qml.PauliY(1))),
+        qml.prod(qml.sum(qml.PauliZ(0), qml.PauliY(1)), qml.s_prod(3, qml.PauliX(2))),
+    ],
+)
+def test_meas_expval(shots, ops):
+    """Test that expectation value measurements with `default.clifford` is possible
+    and agrees with `default.qubit`."""
+    dev_c = qml.device("default.clifford", shots=shots, seed=24)
+    dev_q = qml.device("default.qubit")
+
+    def circuit_fn():
+        circuit_1()
+        return qml.expval(ops)
+
+    qnode_clfrd = qml.QNode(circuit_fn, dev_c)
+    qnode_qubit = qml.QNode(circuit_fn, dev_q)
+
+    assert np.allclose(qnode_clfrd(), qnode_qubit(), atol=1e-2 if shots else 1e-8)
```
Nice!

trbromley reviewed (commented):
Thanks @obliviateandsurrender! No blockers from my side, will be exciting to get this in and have `default.clifford` be more powerful.

trbromley commented:
@obliviateandsurrender I can't remember if you said we need a follow up PR for Pauli noise?

timmysilv commented on tests/devices/test_default_clifford.py:
```diff
@@ -106,26 +106,214 @@ def circuit_empty():
 
     def circuit_fn():
-        circuit()
-        return qml.density_matrix([0, 1]) if meas_type == "dm" else qml.purity([0, 1])
```
I am specifically concerned about multi-qubit density matrices since the new test is only on one

obliviateandsurrender replied:
I did test mult-qubit ones locally, I will add them.

DSGuala reviewed (commented):
I ran some basic tests with the following:
- `qml.expval` with and without shots
- `qml.var` with and without shots
- `qml.qinfo.transforms.purity`
- `qml.counts`
- `qml.shadow_expval`

Didn't run into any bugs! :100: :100: 
I wasn't sure about "support for providing observables with different measurements operations"

timmysilv reviewed (approved):
🥳 🍰 🥇

soranjh commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +221,50 @@ def circuit():
+        One can compute the analytical probability of each computational basis state using
+        :func:`qml.probs <pennylane.probs>`, which also accepts a wire specification for obtaining marginal
+        probabilities and an observable for obtaining probabilities from the rotated computational basis.
+        As the ``default.clifford`` device supports executingm quantum circuits with a large number of qubits,
```
```suggestion
        As the ``default.clifford`` device supports executing quantum circuits with a large number of qubits,
```

soranjh commented on pennylane/devices/default_clifford.py:
```diff
@@ -531,46 +768,238 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
-        if circuit.wires != meas_op.wires:
-            raise NotImplementedError(
-                "default.clifford doesn't support measuring the purity of a subset of wires at the moment."
+    def _measure_stabilizer_vn_entropy(stabilizer, wires, log_base=None):
```
Does stim have any similar method?

obliviateandsurrender replied:
No, stim doesn't have a similar method. However, there is a discussion regarding entropy [here](https://quantumcomputing.stackexchange.com/questions/16718/measuring-entanglement-entropy-using-a-stabilizer-circuit-simulator) essentially based on the A.1.d of this [paper](https://arxiv.org/pdf/1901.08092.pdf) I use.

albi3ro commented on pennylane/operation.py:
```diff
@@ -3036,6 +3037,22 @@ def convert_to_opmath(op):
+    try:
+        enable_new_opmath()
+        yield
+    except Exception as e:
+        raise e
```
```suggestion
```

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +221,50 @@ def circuit():
+            import numpy as np
+
+            dev = qml.device("default.clifford")
+            dev.probability_target = np.array([[0, 0], [1, 0]])
```
A couple questions here.

* **What exactly do these numbers mean?** I'm a little confused by what the numbers actually are

* **Why can't we infer this from the measurements?** In your below example, the probability is on wires 0 and 1.  If the issue is requesting the probabilities on too many wires, can we just raise an error and say "no probabilities on 30+ different wires"?

* Why not specify on initialization? `device('default.qubit', probability_target=probability_target)`?

trbromley replied:
+1 with @albi3ro's comment. Maybe we have a discussion of what we want to do here.

Also note that you can already use `qml.expval(qml.Projector())` for finding specific probabilities.

obliviateandsurrender replied:
> * **What exactly do these numbers mean?** I'm a little confused by what the numbers actually are

These refer to the computational basis states. For example, for a two-qubit system we would have four basis states - `|00>, |10>, |01>, |11>` which would correspond to `[0, 0], [1, 0], [0, 1], [1, 1]`.

> * **Why can't we infer this from the measurements?** In your below example, the probability is on wires 0 and 1.  If the issue is requesting the probabilities on too many wires, can we just raise an error and say "no probabilities on 30+ different wires"?

I did think about it. Added support for it now and restricted qubits size to `24`. 

> * Why not specify on initialization? `device('default.qubit', probability_target=probability_target)`?

Added a keyword argument for specifying it during initialization.

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -254,9 +309,10 @@ def __init__(
 
-        seed = np.random.randint(0, high=10000000) if seed == "global" else seed
-        self._rng = np.random.default_rng(seed)
+        self._seed = np.random.randint(0, high=10000000) if seed == "global" else seed
```
Why a seed and an rng?

obliviateandsurrender replied:
We use stim's `sampler` method for generating samples in the finite shot case, which accepts `seed` and not an `rng`. About the `rng`, I was using it for generating `seed` in the multiprocessing case.

albi3ro replied:
If we are feeding the same seed to the sampler each time, will the samples be identical each time we execute?

obliviateandsurrender replied:
``` py
@qml.qnode(qml.device('default.clifford', shots=50, seed=123455))
def circuit():
    qml.Hadamard(0)
    qml.S(0)
    return qml.expval(qml.PauliZ(0))

circuit(), circuit(), circuit()
```
Tried to fix this. It should be fine now. Above circuit would now give - 

``` py
(tensor(-0.28, requires_grad=True),
 tensor(-0.12, requires_grad=True),
 tensor(0., requires_grad=True))
```

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -460,57 +519,239 @@ def simulate(
+
+        # Case for higher arithmetic depth for prod-type observables
+        if meas_obs.arithmetic_depth > 1 and isinstance(meas_obs, (Prod, SProd)):
+            with qml.operation.use_new_opmath():
+                meas_obs_simp = meas_obs.simplify()
```
Why the need to call simplify inside the `use_new_opmath()` context?  This sounds like a bug we might need to fix.

timmysilv replied:
I don't think we need to anymore, this code pre-dates the fixes I pushed

obliviateandsurrender replied:
Removed this!

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -460,57 +519,239 @@ def simulate(
 
     @staticmethod
-    def pl_to_stim(op):
+    def _pl_to_stim(op):
```
If it's a private static method, does it really need to live on the class? Or could it just be a private helper function at the top of the file?

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -460,57 +519,239 @@ def simulate(
+                results.append(meas.process_samples(samples=samples, wire_order=wire_order))
+        return results
+
+    def measure_analytical(self, circuit, stim_circuit, tableau_simulator, global_phase, stim):
```
Instead of having to pass around the package as a keyword argument, maybe we could follow how we handle importing matplotlib.  In `MPLDrawer.py`, we import matplotlib globally at the top, but defer raising an import error until we try and initialize the class: 

https://github.com/PennyLaneAI/pennylane/blob/321ad706c219ad689de4ee163cf8c37ef017e081/pennylane/drawer/mpldrawer.py#L22

That way we can still have the lazy error, but the package can be global as expected.

albi3ro reviewed (commented):
If `default.clifford` does not support `Projector`, `Hadamard`, or `Hermitian`, could we remove them from the set of supported observables?

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -460,57 +519,239 @@ def simulate(
+        results = []
+        for meas in circuit.measurements:
+            # Computing density matrix via tableaus
+            if isinstance(meas, DensityMatrixMP):  # do first because it is a child of StateMP
```
What would think of moving this logic to either a:

* single dispatch function with signature `(measurement, circuit, stim_circuit, tableau_simulator, global_phase)`

or 

* dictionary mapping type to callable?
```
measurement_map = {
    DensityMatrixMP: self._measure_density_matrix,
    StateMP: self._measure_state,
    ExpectationMP: self._measure_expecation,
    ...
}
results = []
for m in circuit.measurements:
     r = measurement_map[type(m)](measurement, circuit, stim_circuit, tableau_simulator, global_phase)
     results.append(r)
return tuple(results)
```

The pattern of extended `elif`'s can usually be made a lot more readable and maintainable by adhering to a fixed call signature and making the functions interchangeable.  Otherwise, it can be difficult to remember which function had which call signature.

obliviateandsurrender replied:
I did think about it but found it hard to come up with a single dispatch signature, so I had to resort to use the `elif`'s structure. However, I finally added dictionary mapping for stochastic and analytic measurements, as you suggested.

trbromley commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +221,50 @@ def circuit():
+        :title: Probabilities for Basis States
+        :href: clifford-probabilities
+
+        One can compute the analytical probability of each computational basis state using
+        :func:`qml.probs <pennylane.probs>`, which also accepts a wire specification for obtaining marginal
+        probabilities and an observable for obtaining probabilities from the rotated computational basis.
```
```suggestion
```
May not need this info.

trbromley commented on pennylane/devices/default_clifford.py:
```diff
@@ -531,46 +772,238 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
-            raise NotImplementedError(
-                "default.clifford doesn't support measuring the purity of a subset of wires at the moment."
+    def _measure_stabilizer_vn_entropy(stabilizer, wires, log_base=None):
+        r"""Computes the Rényi entanglement entropy using stabilizer information.
```
Is this Renyi or VN?

obliviateandsurrender replied:
They are equivalent in the case of stabilizer circuits.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -460,57 +519,239 @@ def simulate(
-    def measure(self, circuit, tableau_simulator, global_phase, stim):
-        """Given a circuit, compute and return the measurement results."""
+    # pylint:disable=too-many-return-statements
+    def _convert_op_to_linear_comb(self, meas_obs, coeffs, paulis):
```
Thanks, I wasn't aware of `pauli_rep`. It really helped simplify the function.

obliviateandsurrender commented:
> If `default.clifford` does not support `Projector`, `Hadamard`, or `Hermitian`, could we remove them from the set of supported observables?

Added support for `Projector` and `Hermitian`. Removed `Hadamard` from the list as it would require `T` gates for rotating the basis.

Qottmann commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +254,53 @@ def circuit():
+        :title: Probabilities for Basis States
+        :href: clifford-probabilities
+
+        As the ``default.clifford`` device supports executing quantum circuits with a large number of qubits,
+        we restrict the ability to compute the ``analytical`` probabilities for `all` computational basis
+        states at once to maintain computational efficiency when system size scales beyond ``24`` qubits.
```
Could you make it a bit more explicit why probabilities are special in that even though we can efficiently simulate the circuit, there are still exponentially many of them so that is why we need to restrict them

Qottmann commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +254,53 @@ def circuit():
+            import pennylane as qml
+            import numpy as np
+
+            basis_states = np.array([[0, 0], [1, 0]])
```
Not sure this would be clear to everyone from context. I think it would be good to have in the text somewhere that the states we are targeting here are $|00\rangle$ and $|10\rangle$, then it should be clear from context.

Qottmann commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +254,53 @@ def circuit():
+            basis_states = np.array([[0, 0], [1, 0]])
+            dev = qml.device("default.clifford", target_states = basis_states)
+
+        After doing this, one can simply use the :func:`qml.probs <pennylane.probs>` with its usual arguments
+        and probabilities for the specified target states would be computed and returned.
```
Could also expand just a little bit what those usual arguments are, i.e. what is up with those operators?

Qottmann commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +254,53 @@ def circuit():
+                    qml.PauliX(w[0])
+                    qml.PauliY(w[1])
+                    qml.PauliZ(w[2])
+                return qml.probs(op = qml.PauliX(0) @ qml.PauliY(1))
```
After explaining about the target states this may be a bit confusing. Like what are those operators doing here? Can you give a mathematical expression for what it is we are computing, I think that would be very helpful to the reader.

Qottmann commented on pennylane/devices/default_clifford.py:
```diff
@@ -245,18 +336,28 @@ def __init__(
         seed="global",
         max_workers=None,
     ) -> None:
+        if not has_stim:
+            raise ImportError(
+                "This feature requires stim, a fast stabilizer circuit simulator. "
+                "It can be installed with:\n\npip install stim"
+            )
```
Maybe a bit late to the game, but why do we call this device `default.clifford` if it is a PennyLane interface to `stim`? _Could_ be perceived as adorning ourselves with borrowed plumes, not sure, the device doc first line states and links stim so could be fine 🤔 CC @trbromley

Qottmann commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +686,298 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+    def _measure_stabilizer_entropy(stabilizer, wires, log_base=None):
+        r"""Computes the Rényi entanglement entropy using stabilizer information.
+
+        Computes the Rényi entanglement entropy :math:`S_A` for a subsytem described by
+        :math:`A`, :math:`S_A = \text{rank}(\text{proj}_A {\mathcal{S}}) - |\mathcal{S}|`,
+        where :math:`\mathcal{S}` is the stabilizer group for the system using the theory
+        described in `arXiv:1901.08092 <https://arxiv.org/abs/1901.08092>`_.
```
Didnt know that, pretty cool! Might be worth to make this user-facing somewhere!

Qottmann reviewed (commented):
Very cool work @obliviateandsurrender 👌 

Left some high level comments mostly on docs in a first pass :)

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -452,65 +555,135 @@ def simulate(
-        """Given a circuit, compute and return the measurement results."""
+    def _measure_observable_sample(meas_obs, stim_circuit, shots, sample_seed):
+        """Compute sample output from a stim circuit for a given Pauli observable"""
+        meas_dict = {"X": "MX", "Y": "MY", "Z": "MZ", "_": "M"}
```
what is `"_"`?

obliviateandsurrender replied:
"_" refers to Identity (kind of placeholder) used by the stim for its Pauli string representation.

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +692,322 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+
+        return qml.math.dot(coeffs, expecs)
+
+    def _measure_variance(self, meas, tableau_simulator, **kwargs):
+        """Measure the variance with respect to the state of simulator device."""
+        meas_obs, _ = qml.operation.convert_to_opmath(meas.obs), kwargs
```
your pylint solution made me smile but this would do just fine and probably confuse people less
```suggestion
    def _measure_variance(self, meas, tableau_simulator, **_):
        """Measure the variance with respect to the state of simulator device."""
        meas_obs = qml.operation.convert_to_opmath(meas.obs)
```

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +692,322 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+            raise ValueError("Cannot set an empty list of target states.")
+
+        if len(meas_wires) < tgt_states.shape[1]:
+            tgt_states = []
+            for state in self._prob_states:
+                if list(state[meas_wires]) not in tgt_states:
+                    tgt_states.append(list(state[meas_wires]))
+            tgt_states = np.array(tgt_states)
```
what is the type of `state[meas_wires]` before casting to list? can we use `set()` construction instead to speed this up? no worries if it isn't trivial work

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +692,322 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+            prob_sim = circuit_simulator.copy()
+            for idx, wire in enumerate(meas_wires):
+                expectation = prob_sim.peek_z(wire)
+                # (Eig --> Res) | -1 --> 1 | 1 --> 0 | 0 --> 0 / 1 |
+                outcome = int(0.5 * (1 - expectation))
```
is the `0 --> 0 / 1` randomness from rounding `0.5`, and if so what's the consequence of that?

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +692,322 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+    def _sample_expectation(self, meas, stim_circuit, shots, seed):
+        """Measure the expectation value with respect to samples from simulator device."""
+        # Get the observable for the expectation value measurement
+        meas_op = meas.obs
+        if meas_op is None:  # pragma: no cover
+            meas_op = qml.prod(
+                *[
+                    qml.PauliZ(idx)
+                    for idx in (meas.wires if meas.wires else range(stim_circuit.num_qubits))
+                ]
+            )
```
shouldn't be possible, expval needs an observable. also, you still use `meas.obs` in the next line (which is fine), so we can just not define `meas_op` here
```suggestion
```

obliviateandsurrender replied:
I had added them to ensure operability if someone (possibly a `dev`) uses these internal methods directly for their workflow with `ExpectationMP` directly. 😓

timmysilv commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +692,322 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+        """Measure the variance with respect to samples from simulator device."""
+        # Get the observable for the expectation value measurement
+        meas_op = meas.obs
+        if meas_op is None:  # pragma: no cover
+            return 1 - self._sample_expectation(meas, stim_circuit, shots, seed) ** 2
+
```
same concern as expval
```suggestion
```

timmysilv commented on pennylane/operation.py:
```diff
@@ -3039,6 +3040,20 @@ def convert_to_opmath(op):
 
 
+@contextmanager
+def use_new_opmath():
```
i'm fine with leaving this but i'm not sure if anyone wants it anymore... @trbromley @Qottmann ?

trbromley replied:
I don't think this context manager is used outside of this PR.

timmysilv reviewed (commented):
new changes also look great! left a few comments on things but nothing serious.

on the topic of `probability_target` - do we need to provide the option to re-define it after device construction/will users want to change it often? I believe a concern @albi3ro had was making the device as immutable as possible. If we don't need to offer that, then we should remove the setter. If we do need/want to offer it, maybe we should make it a runtime option instead of a device configuration thing. can be clumsy, so we can work together on making that happen if necessary

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +251,64 @@ def circuit():
+
+            import pennylane as qml
+            import numpy as np
+            dev = qml.device("default.clifford", target_states = basis_states)
```
Mind defining `basis_states`?

obliviateandsurrender replied:
Sorry, I forgot to remove it from here we have now removed the `target_states` keyword.

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -99,27 +115,52 @@ def observable_stopping_condition(obs: qml.operation.Operator) -> bool:
         wires (int, Iterable[Number, str]): Number of wires present on the device, or iterable that
             contains unique labels for the wires as numbers (i.e., ``[-1, 0, 2]``) or strings
-            (``['ancilla', 'q1', 'q2']``). Default ``None`` if not specified.
+            (``['aux_wire', 'q1', 'q2']``). Default ``None`` if not specified.
```
Mind adding `target_states` to the list of arugments?

trbromley commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +251,64 @@ def circuit():
+        :title: Probabilities for Basis States
+        :href: clifford-probabilities
+
+        As the ``default.clifford`` device supports executing quantum circuits with a large number of qubits,
+        we restrict the ability to compute the ``analytical`` probabilities for `all` computational basis
+        states at once to maintain computational efficiency when system size scales beyond ``20`` qubits.
```
Is this still a hardcoded restriction? I don't think we need to check this, similar to how we don't check if a user tries to run a 100 qubit circuit in `default.qubit`.

trbromley commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +251,64 @@ def circuit():
+                    qml.CNOT(wires=[idx, idx+1])
+                return qml.expval(qml.Projector(state, wires=range(num_wires)))
+
+        >>> basis_states = np.array([[0, 0, 0, 0], [0, 1, 0, 1], [1, 0, 1, 0]])
+        >>> circuit(basis_states[0])
+        tensor(0.5, requires_grad=True)
+        >>> circuit(basis_states[1])
+        tensor(0.0, requires_grad=True)
+        >>> circuit(basis_states[2])
+        tensor(0.0, requires_grad=True)
```
Nice!

trbromley commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +251,64 @@ def circuit():
+        >>> circuit(basis_states[2])
+        tensor(0.0, requires_grad=True)
+
+        .. note::
+
+            If there's a mismatch in the number of ``wires`` for the provided target state(s) and
+            the observable, then the `marginal` probabilities will be returned for the computational
+            basis states for the subspace built using the minimum number of wires among the two.
```
Do we still need this note?

obliviateandsurrender replied:
I think we can remove it now!

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -99,27 +115,52 @@ def observable_stopping_condition(obs: qml.operation.Operator) -> bool:
+            f"default.clifford doesn't support expectation value calculation with {type(meas_op).__name__} at the moment."
+        )
+
+    coeffs, ops = meas_rep.hamiltonian(wire_order=meas_obs.wires).terms()
+
+    # Build the Pauli representation for stim
+    paulis = []
+    for op in ops:
+        op_names = op.name if isinstance(op, qml.operation.Tensor) else [op.name]
+        paulis.append(("".join([_GATE_OPERATIONS[name] for name in op_names]), op.wires))
 
+    return coeffs, paulis
```
Just so we don't need to come back to this next month.
```suggestion
    coeffs = meas_rep.values()
    paulis = [("".join(pw.values()), tuple(pw.keys()) ) for pw in op.pauli_rep]
    return coeffs, paulis
```

obliviateandsurrender replied:
Added it after some modification for handling Identities which go missing in `pauli_rep`.

albi3ro commented on pennylane/devices/default_clifford.py:
```diff
@@ -347,16 +449,19 @@ def execute(
         max_workers = execution_config.device_options.get("max_workers", self._max_workers)
         if max_workers is None:
-            results = tuple(self.simulate(c, debugger=self._debugger) for c in circuits)
+            seeds = self._rng.integers(2**31 - 1, size=len(circuits))
```
Thanks 🎉

albi3ro reviewed (approved):
Thanks @obliviateandsurrender  🎉 Looking good to go from me :)

trbromley commented on pennylane/devices/default_clifford.py:
```diff
@@ -210,6 +249,60 @@ def circuit():
+        >>> circuit()
+        tensor([0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5], requires_grad=True)
+
+        Once above the limit, one can use :mod:`qml.expval <pennylane.expval>` compute the expectation value
```
```suggestion
        Once above the limit, one can use :mod:`qml.expval <pennylane.expval>` to compute the expectation value
```

trbromley reviewed (approved):
Thanks, approved from the docs perspective!

soranjh commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +685,306 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+        )
+
+    @staticmethod
+    def _measure_single_sample(stim_ct, meas_ops, meas_idx, meas_wire):
```
Does it merit a dedicated function? Could this be also `_measure_observable_sample`?

obliviateandsurrender replied:
Yes, this is specialized for efficiently calculating shadows. I would like to keep it separate.

soranjh commented on tests/devices/test_default_clifford.py:
```diff
@@ -94,38 +93,276 @@ def circuit_fn():
+    """Test if probabilities are returned in the clifford device."""
+
+    dev_c = qml.device("default.clifford", tableau=tableau, shots=shots, seed=24)
+    dev_q = qml.device("default.qubit")
```
No need for shots here?

obliviateandsurrender replied:
I am testing the sampling results from `default.clifford` against the analytic solution from `default.qubit`, so no shots are required for the latter.

soranjh commented on tests/devices/test_default_clifford.py:
```diff
@@ -94,38 +93,276 @@ def circuit_fn():
+    qnode_qubit = qml.QNode(circuit_fn, dev_q)
+
+    gotten_probs, target_probs = qnode_clfrd(), qnode_qubit()
+    if gotten_probs.shape[-1] != target_probs.shape[-1]:
+        target_probs = qml.math.sum(target_probs.reshape((2,) * len(ops.wires)), axis=-1).reshape(
+            1, -1
+        )
```
Why this is needed?

obliviateandsurrender replied:
It's no longer needed, removed it!

soranjh commented on tests/devices/test_default_clifford.py:
```diff
@@ -94,38 +93,276 @@ def circuit_fn():
+            qml.PauliX(wire)
+            qml.PauliY(wire)
+            qml.PauliZ(wire)
+        return qml.probs(wires=range(1))
```
Is `range(1)` intentional?

obliviateandsurrender replied:
Yes.

soranjh commented on tests/devices/test_default_clifford.py:
```diff
@@ -94,38 +93,276 @@ def circuit_fn():
+    assert list(counts_clfrd.keys()) == list(counts_qubit.keys())
+
+    for k1, k2 in zip(counts_clfrd, counts_clfrd):
+        assert qml.math.abs(counts_clfrd[k1] - counts_qubit[k2]) / shots < 5.0
```
What is the 5 here?

obliviateandsurrender replied:
Oops, it was supposed to be % threshold, fixed it.

soranjh reviewed (changes requested):
Thanks @obliviateandsurrender, looks good overall. Left some comments and questions. We can also discuss the logic behind `_measure_probability` offline.

obliviateandsurrender commented on .gitignore:
```diff
@@ -8,7 +8,7 @@ qit.egg-info
 *.ipynb_checkpoints
 *.egg-info*
-.coverage
+.coverage*
```
It is intentional. Prevents some extra coverage files from being tracked when using `--cov` in `pytest`.

soranjh commented on tests/devices/test_default_clifford.py:
```diff
@@ -94,38 +93,276 @@ def circuit_fn():
+
+    basis_states = np.array([[0, 0], [1, 0], [0, 1], [1, 1]])
+    probs1 = [qnode_clfrd(state) for state in basis_states]
+    probs2 = [1.0, 0.0, 0.0, 0.0]
+    assert qml.math.allclose(probs1, probs2)
```
[non-blocking] This does not seem like a powerful test.

obliviateandsurrender replied:
Although I think `test_meas_probs` (the test above it) sufficiently tests the probabilities, I have made this one stronger as well.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -99,27 +115,50 @@ def observable_stopping_condition(obs: qml.operation.Operator) -> bool:
     return obs.name in _MEAS_OBSERVABLES
 
 
-def _import_stim():
```
This was removed to use an approach suggested [here](https://github.com/PennyLaneAI/pennylane/pull/4954/files/5331f988a30677af03cc71c53e46b35c3a271ec8#r1470195801), in order to not having to move around `stim` across multiple functions.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -45,6 +57,12 @@
+has_stim = True
+try:
+    import stim
+except (ModuleNotFoundError, ImportError) as import_error:  # pragma: no cover
```
I am already testing the import [here](https://github.com/PennyLaneAI/pennylane/blob/979db26936bf7076f2c6679e796f3d4a0fc7b759/tests/devices/test_default_clifford.py#L678) with `monkeypatch`. For testing the particular line, one will have to reimport the pennylane module in a function call, which might interfere with other tests, so skipped it.

albi3ro replied:
My experience with trying to patch imports is that it often messes something up, and as it doesn't always clean up after itself sufficiently.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -99,27 +115,50 @@ def observable_stopping_condition(obs: qml.operation.Operator) -> bool:
+    return stim_op, " ".join(map(str, op.wires))
+
+
+def _convert_op_to_linear_comb(meas_op):
```
Yes, it is called for all the observables listed in the `_MEAS_OBSERVABLES` dictionary, which includes - any single qubit or tensored Pauli observable, Basis state projectors, Hermitian matrices, Hamiltonians and the new operator arithmetic objects.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -452,65 +549,134 @@ def simulate(
+        return meas_results[0] if len(meas_results) == 1 else tuple(meas_results)
 
     @staticmethod
-    def pl_to_stim(op):
```
Yep, just shifted it outside the device class and made it a private [method](https://github.com/PennyLaneAI/pennylane/blob/979db26936bf7076f2c6679e796f3d4a0fc7b759/pennylane/devices/default_clifford.py#L132).

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -452,65 +549,134 @@ def simulate(
+            if measurement_func is not None:
+                res = measurement_func(meas, stim_circuit, shots=num_shots, seed=sample_seed)
+            else:
+                meas_wires = meas.wires if meas.wires else range(stim_circuit.num_qubits)
+                meas_op = meas.obs or qml.prod(*[qml.PauliZ(idx) for idx in meas_wires])
+                samples = qml.math.array(
+                    [
+                        self._measure_observable_sample(
+                            meas_op, stim_circuit, num_shots, sample_seed
+                        )[0]
+                    ]
+                )
+                wire_order = {wire: idx for idx, wire in enumerate(meas.wires)}
+                res = (
+                    qml.math.squeeze(samples[0], axis=0)
+                    if isinstance(meas, SampleMP)
+                    else meas.process_samples(samples=samples, wire_order=wire_order)
+                )
```
So, for the sampling measurements that don't require the specialized measurement method (mentioned in the `measurement_map` dictionary), I simply compute the `samples` based on the observable given as part of measurement (if not, we use the computational basis as a fallback) from `stim` and then use the existing PL functionality with `prcoess_samples` to do the required calculation. This allows me to re-invent the wheel for the measurements that can be directly computed from the samples.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -452,65 +549,134 @@ def simulate(
+        wires = kwargs.get("circuit").wires
+        global_phase = kwargs.get("global_phase", qml.GlobalPhase(0.0))
         if self._tableau:
+            # Stack according to Sec. III, arXiv:0406196 (2008)
```
`Stim` returns the tableau as a list of multiple arrays. I have stacked them together as a complete *stabilizer* `Tableau` matrix as described/introduced in this particular paper by Gottesman and Aaronson.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +685,306 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+        Returns:
+            (float): entanglement entropy of the subsystem
+        """
+        # Get the number of qubits for the system
+        num_qubits = qml.math.shape(stabilizer)[0]
+
+        # Von Neumann entropy of a stabilizer state is zero
+        if len(wires) == num_qubits:
+            return 0.0
```
This means that a stabilizer state will always be a pure state, and the state of the devices will be a stabilizer state whenever we are not tracing over any qubits (using all the wires).

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +685,306 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
-        # Add support for more case when the time is right
-        raise NotImplementedError(
-            f"default.clifford doesn't support expectation value calculation with {type(meas_op.obs)} at the moment."
+        # Use the Eq. A17 in arXiv:1901.08092 for entropy calculation
```
We mention the formulation we use in the docstring. I don't think we require a derivation, as interested readers can simply look at the paper I cite. Tweaked the docs to reflect finding the required section better.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +685,306 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+        """Measure the expectation value with respect to samples from simulator device."""
+        # Get the observable for the expectation value measurement
+        meas_op = meas.obs
+        if isinstance(meas_op, BasisStateProjector):
+            stim_circ = stim_circuit.copy()
+            stim_circ.append_from_stim_program_text("M " + " ".join(map(str, meas_op.wires)))
+            sampler = stim_circ.compile_sampler(seed=seed)
+            samples = qml.math.array(sampler.sample(shots=shots), dtype=int)
+            matches = np.where((samples == meas_op.data[0]).all(axis=1))[0]
+            return len(matches) / shots
```
Great suggestion; shifted the sample computation part to there.

soranjh commented on pennylane/devices/default_clifford.py:
```diff
@@ -99,27 +116,50 @@ def observable_stopping_condition(obs: qml.operation.Operator) -> bool:
+    return stim_op, " ".join(map(str, op.wires))
+
+
+def _convert_op_to_linear_comb(meas_op):
```
It seems that you have `pl_to_stim` for converting operators and  `_convert_op_to_linear_comb` for converting observables? If it is the case, I suggest using a more intuitive name here. Even private functions should have intuitive/consistent names and documentations for easier maintenance and improvement in the future.

obliviateandsurrender replied:
I originally had a bigger conversion routine for handling all types and classes of PennyLane observables, and so this method name has been a remnant of it. As requested, I have changed the name for both of the functions and attempted to make it more intuitive, and I feel docstrings summarizes their functionality well enough. However, if there are any other major concerns regarding docstrings and function names as such, I would be happy to deal with them and improve upon them in a separate PR.

soranjh commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +698,310 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+    def _measure_probability(self, meas, _, **kwargs):
+        """Measure the probability of each computational basis state.
+
+        Computes the probability for each computational basis state vector. This is done by
+        building the complete basis using a vectorized integer to bit-array transformation.
+        Then, for each basis state, one iterates over each of the measured qubits, and checks
+        the possibility of that state and performs post-selection. If an impossible state is
+        encountered, the corresponding branches are tracked and marked in a visited-array.
```
This is not clear. Having no explanation is better than having one that is not helpful.

obliviateandsurrender replied:
I have replaced this summary with the `pseudocode` for the algorithm I implement, which now breaks down the algorithm into five sequential parts. I am already specifying most of the workings, including the complexity as comments with the function definition, which is why I wanted to keep the docstring more succinct and manageable.

obliviateandsurrender commented on pennylane/devices/default_clifford.py:
```diff
@@ -519,58 +698,310 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+        r"""Computes the Rényi entanglement entropy using stabilizer information.
+
+        Computes the Rényi entanglement entropy :math:`S_A` for a subsytem described by
+        :math:`A`, :math:`S_A = \text{rank}(\text{proj}_A {\mathcal{S}}) - |\mathcal{S}|`,
```
It was a typo, corrected it.
